### ISSUE TITLE
Different float/half overload behaviour than fxc

### ISSUE THREAD

**Issue by simontaylor81** [Microsoft/DirectXShaderCompiler#1397]
dxc exhibits different behaviour to fxc when overloading on half vs. float. fxc will happily do this, but dxc seems to effectively typedef half to float, so throws a redefinition error.

Repro:
```
half func(half value1) { return 1; }
float func(float value1) { return 2; }

float4 PSMain() : SV_TARGET
{
	return func(0.0f);
}
```

Also here: http://shader-playground.timjones.io/4ff8550e2014e051980461f1117a719c

Error from dxc:
```
D:\local\Temp\f4e285aa-ff0d-48db-a9ec-3a054fa77cfe.hlsl:2:7: error: redefinition of 'func'
float func(float value1) { return 2; }
      ^
D:\local\Temp\f4e285aa-ff0d-48db-a9ec-3a054fa77cfe.hlsl:1:6: note: previous definition is here
half func(half value1) { return 1; }
     ^
```

Note: I am in know way suggesting that dxc's behaviour is _worse_! We found this because of some existing code in our codebase, but I consider it Bad Code that needs to be fixed. I'm raising this merely to provide awareness in case it is not already known. I personally would be happy for this to be a documented departure from fxc behaviour.

**Comment by hekota** [comment]
In dxc you need to use -enable-16bit-types flag if you want to preserve the difference between half and float.

**Comment by simontaylor81** [comment]
Ah, fair enough! We'll just fix our code :)

**Comment by tex3d** [comment]
Just FYI @simontaylor81: This departure from fxc was unintentional, and has been fixed since this issue was closed (#1492).

The -enable-16bit-types flag is not quite what you're looking for here, since it will not give you the same result as fxc.  In fxc, when compiling to DX10+ targets, half means the same thing as float, but is tracked as a separate basic type.  For certain DX9 targets, it did mean something different (partial precision), but that meaning was left behind when moving to DX10 shader targets.

For DX10+ targets, the replacements for half (added later) were some min-precision types (`min16float`, for example).  If you use the -enable-16bit-types flag, you will get strict 16-bit types, with the corresponding shader model (6.2) and feature flag requirements that must be supported by your device.  `half` will map to `float16_t`.  This will also change min-precision types to be strict 16-bit precision, since these features are mutually exclusive in DXIL.

**Comment by simontaylor81** [comment]
Ok, thanks for the heads up! The code that triggered this was dx9 legacy, so it was a good excuse to clean it up :)
________________________________
